[docs][Base] useAutocomplete demos & docs (#37029)
mui/material-ui: docs/data/base/components/autocomplete/autocomplete.md

To render the listbox in Base UI's Popper, the `ref`s must be merged as follows:

```jsx
import useAutocomplete from '@mui/base/useAutocomplete';
import Popper from '@mui/base/Popper';
import { unstable_useForkRef as useForkRef } from '@mui/utils';

export default function App(props) {
  const {
    getRootProps,
    getInputProps,
    getListboxProps,
    getOptionProps,
    popupOpen,
    anchorEl,
    setAnchorEl,
    groupedOptions,
  } = useAutocomplete(props);

  const rootRef = useForkRef(ref, setAnchorEl);

  return (
    <>
      <div {...getRootProps()} ref={rootRef}>
        <input {...getInputProps()} />
      </div>
      {anchorEl && (
        <Popper open={popupOpen} anchorEl={anchorEl}>
          {groupedOptions.length > 0 && (
            <ul {...getListboxProps()}>
              {groupedOptions.map((option, index) => (
                <li {...getOptionProps({ option, index })}>{option.label}</li>
              ))}
            </ul>
          )}
        </Popper>
      )}
    </>
  );
}
```

Here's a complete demo that renders the listbox inside a Popper:

> {{"demo": "UseAutocompletePopper.js", "hideToolbar": true}}

Author: It's annoying how you have to scroll a little bit to see the important part of the code, that's why I hid it here and manually put the whole snippet above ~

> {{"demo": "UseAutocompletePopper.js", "hideToolbar": true}}

Author: 👌 I actually just meant to hide the preview

waiAria: https://www.w3.org/WAI/ARIA/apg/patterns/combobox/
---

# Autocomplete

<p class="description">The Autocomplete component is a text input enhanced by a panel of suggested options.</p>

{{"component": "modules/components/ComponentLinkHeader.js", "design": false}}

{{"component": "modules/components/ComponentPageTabs.js"}}

## Introduction

The autocomplete component is an enhanced text input that shows a list of suggested options as users type and lets them select an option from the list. It implements the WAI-ARIA Combobox pattern, and is typically used to assist users complete a form inputs or search queries faster.

{{"demo": "AutocompleteIntroduction.js", "defaultCodeOpen": false, "bg": "gradient"}}

:::warning
Material UI and Joy UI have Autocomplete components that are built using the useAutocomplete hook, with many features that are not demo-ed here yet.

To learn more about implementing a custom Autocomplete, you can explore the useAutocomplete API [docs](/base/react-autocomplete/hooks-api/), or reference the Material UI and Joy UI implementations:

- [Material UI Autocomplete](/material-ui/react-autocomplete/)
- [Joy UI Autocomplete](/joy-ui/react-autocomplete/)

> :::

Author: @samuelsycamore I think the message in this callout is a little clumsy, any suggestions? 😅

Reviewer: Looks pretty good to me! Here's how I'd phrase that first part.

```suggestion
:::warning
Material UI and Joy UI have Autocomplete components that are built using the `useAutocomplete` hook, and they include many features not yet described here.

To learn more about implementing a custom Autocomplete, you can explore the [`useAutocomplete` API docs](/base/react-autocomplete/hooks-api/), or reference the Material UI and Joy UI implementations:

- [Material UI Autocomplete](/material-ui/react-autocomplete/)
- [Joy UI Autocomplete](/joy-ui/react-autocomplete/)

:::
```

      { label: '12 Angry Men', year: 1957 },
      { label: "Schindler's List", year: 1993 },
    ],
    getOptionLabel: (option) => option.label,
  });

  return (
    <>
      <div {...getRootProps()}>
        <input {...getInputProps()} />
      </div>
      {groupedOptions.length > 0 && (
        <ul {...getListboxProps()}>
          {groupedOptions.map((option, index) => (
            <li {...getOptionProps({ option, index })}>{option.label}</li>
          ))}
        </ul>
      )}
    </>
  );
}
```

## Basics

The useAutocomplete hook requires a list of `options` to be displayed when the textbox receives focus. The value must be chosen from a predefined set of values.

The following demo shows how to create a simple combobox, apply some styling, and write the selected value to a state variable using the `onChange` prop:

{{"demo": "UseAutocomplete.js"}}

## Customization

### Rendering options

By default, the `options` prop accepts an array of `string`s or `{ label: string }`:

```js
const options = [
  { label: 'The Godfather', id: 1 },
  { label: 'Pulp Fiction', id: 2 },
];
// or
const options = ['The Godfather', 'Pulp Fiction'];
```

> If you need to use a different struction for options, you must provide a function to the `getOptionLabel` prop that resolves each option to a unique value.

Author: I modified this bit compared to the other docs (Material, Joy) because it kind of understates the importance of `getOptionLabel`.

The API docs just say “Used to determine the string value for a given option”, but it seems the hook completely relies on `getOptionLabel` to tell different options from each other

I noticed issues kind of related to this:
- https://github.com/mui/material-ui/issues/36897
- https://github.com/mui/material-ui/issues/26492
